Add gomc parser

msoroush opened the pull request:
Add GOMC parser to alchemlyb. Energy output file for Free Energy Calculation is similar to GROMCAS. This parser, reads, Total energy, PV, derivative of energy (for current lambda state), and change of energy (between current lambda state and all other lambda states).

dotsdl commented:
@msoroush thanks for opening this PR! I've scheduled it for review on 2019.05.05.

orbeckst commented:
You'll need tests.

For tests you need to add a data set to [alchemtest](https://github.com/alchemistry/alchemtest), so open an issue/PR over there, please. See https://github.com/alchemistry/alchemtest/wiki/contributing

msoroush commented:
> You'll need tests.
> 
> For tests you need to add a data set to [alchemtest](https://github.com/alchemistry/alchemtest), so open an issue/PR over there, please. See https://github.com/alchemistry/alchemtest/wiki/contributing

I am still validating my implementation in GOMC. Once it validate it, I will create the test. I might cancel this PR because I noticed that slicing function does not work index name other than `time`.

orbeckst commented:
See https://github.com/alchemistry/alchemlyb/issues/79#issuecomment-489168116

orbeckst commented:
Btw, you don't have to close the PR. You can keep it open and update it.

Thanks for offering to contribute. As I said, it would be good to have code to deal with MC data. We just have to make sure that the code remains maintainable and one of the key solutions there is to do testing really well. 

Please also be aware that there's no-one being paid to do development on alchemlyb so any code reviews or comments might take some time. We are all way to busy and we try to carve out some time to move this project forward. So please have some patience with us. It's ok to ping someone with their GitHub handle @-mention (eg @orbeckst for me) after a few days if there follow-up seems to be missing.

msoroush commented:
>  Please also be aware that there's no-one being paid to do development on alchemlyb so any code reviews or comments might take some time. We are all way to busy and we try to carve out some time to move this project forward. So please have some patience with us. It's ok to ping someone with their GitHub handle @-mention (eg @orbeckst for me) after a few days if there follow-up seems to be missing.

I will keep it open and work on it. I tried to keep the file format to be similar to GROMACS to make everyones life easy. I did not want to write analysis tools from scratch, so I started to use alchemical-analysis and then alchemlyb. This is a new field for me and I am trying to learn it as fast as possible. Thats why I have too many questions.

dotsdl commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,219 @@
+    return state, lambdas, statevec
+
+
+def _extract_legend(file):
```
Please remove this function, since it is not needed.

msoroush commented:
> You'll need tests.
> 
> For tests you need to add a data set to [alchemtest](https://github.com/alchemistry/alchemtest), so open an issue/PR over there, please. See https://github.com/alchemistry/alchemtest/wiki/contributing

Does it matter which forcefield or water model I am using to generate benzene solvation data set?

orbeckst commented:
No, ideally it should be something publicly available. You should open an issue and PR in [alchemtest](https://github.com/alchemistry/alchemtest) and link it to this one by including  alchemistry/alchemlyb#78 in the description.

orbeckst commented:
See [contributing to alchemtest](https://github.com/alchemistry/alchemtest/wiki/contributing) for guide lines. Ask if you have questions.

msoroush commented:
> can you please add tests for 
> - parsing
 >- calculating free energies: you have both FEP and TI parsers so you should also calculate your 
 solvation free energy with BAR/MBAR and TI

Where should I add test for parsing and calculating free energies?

orbeckst commented:
* Tests for parsing: new file `test_gomc.py` under [src/alchemlyb/tests/parsing](https://github.com/alchemistry/alchemlyb/tree/master/src/alchemlyb/tests/parsing).
* Tests for FEP (BAR/MBAR – your choice): [src/alchemlyb/tests/test_fep_estimators.py](https://github.com/alchemistry/alchemlyb/blob/master/src/alchemlyb/tests/test_fep_estimators.py)
* Tests for TI: [src/alchemlyb/tests/test_ti_estimators.py](https://github.com/alchemistry/alchemlyb/blob/master/src/alchemlyb/tests/test_ti_estimators.py)

Have a look at the existing tests and model yours after them.

msoroush commented:
@orbeckst I added the tests for GOMC parser and free energy estimator (BAR, MBAR, and TI). Please let me know If you want me to add additional test for GOMC parser?

I noticed that I would not load benzene data from GOMC datasets. I fixed the access.py file  https://github.com/alchemistry/alchemtest/pull/36.

msoroush commented:
@orbeckst would you like me to merge the master into my repository, resolve the conflict, and push it back?

dotsdl commented:
@msoroush please do. As for the test failures, they appear to be unrelated to this PR. I'm working on them now.

msoroush commented:
> @msoroush please do. As for the test failures, they appear to be unrelated to this PR. I'm working on them now.

Okay. I think test failure was due to the bug in alchemtest, it would not load the benzene files.

dotsdl commented:
We're also seeing failures due to deprecations in the latest pandas release. We have some duplicate columns in the `alchemtest.gmx` Benzene dataset, and these need to be handled explicitly by our parser. Not something you have to worry about in this PR.

orbeckst commented:
@dotsdl if you have an idea where the "duplicate columns" failures come from could you please raise a separate issue? Thanks.

orbeckst commented:
@msoroush please also add yourself to the new AUTHORS file and add an entry to CHANGES, referencing issue #77 .

dotsdl commented:
When #86 is merged into master, please merge `master` into this branch. That should fix our tests.

dotsdl commented:
@orbeckst #86 is a fix for this duplicate issue.

orbeckst commented:
@msoroush I updated your branch so that you are back in sync with upstream/master (due to PR #86 ) but **it is generally not a good idea to merge other pull-request branches into your own PR branch**. The point of PRs is to check individual code contributions separately. If you add another PR into yours and we decide to change the other PR then you have to also merge these changes. Just keep things separate. If something else needs to be fixed first then this other thing simply has to happen first.

As a rule of thumb: *only merge upstream/master* into your PR branch.

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+
+    Parameters
+    ----------
+    file : str
```
filename

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    file : str
+        Path to free energy file to extract data from.
+    T : float
+        Temperature in Kelvin the simulations sampled.
```
of the simulation sampled

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    return u_k
+
+
+def extract_dHdl(file, T):
```
filename

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+
+    Parameters
+    ----------
+    file : str
```
filename

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    dHdl = df[dHcols]
+
+    # make dimensionless
+    dHdl = beta * dHdl
```
nicer
```python
dHdl *= beta
```

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    return dHdl
+
+
+def _extract_state(file):
```
filename

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+                lambdas = ['Coulomb', 'VDW']
+                statevec = eval(line.strip().split(' = ')[-1])
+                break
+
```
I think this function parses the whole file. Is there a way that you can break once you have all necessary state information?

It's ok not to address this problem right now and leave it as an optimization for later (we have the same problem in other readers, too). But if it is the case that this can be optimized, add a comment saying so.

msoroush replied:
In line 164, we break once we read all necessary state information. In GOMC output format, these information is printed in the first line. So, we never go through the whole file.

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    return state, lambdas, statevec
+
+
+def _extract_dataframe(file):
```
filename

msoroush commented:
@orbeckst do you want me to update [docs/api_proposal.rst](https://github.com/alchemistry/alchemlyb/blob/master/docs/api_proposal.rst) as well?

orbeckst commented:
No need to update `api_proposal` – it's the general ideas and outline. Thanks for checking.

orbeckst commented:
The codecov status is missing because the upload from Travis is failing
```
Error: HTTPSConnectionPool(host='codecov.io', port=443): Max retries exceeded with url: /codecov/v4/raw/2019-07-26/4CA02E04D860053FE833B077F7D9C963/98bf9291d28dc957b5e82407d222dbcd847eb2c6/b3be7e92-5b6c-4e13-a9e6-223a1411d0b2.txt?AWSAccessKeyId=AKIAIHLZSCQCS4WIHD4A&Expires=1564165667&Signature=ouAwWnD%2FfiB%2FP0u7%2Bidshrlg%2Bqk%3D (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb6260d7da0>: Failed to establish a new connection: [Errno 110] Connection timed out'))
```
Not sure why.

orbeckst commented:
@dotsdl can you have a brief look at this PR and if your requirements were addressed? It seems good to go.

dotsdl commented:
I should be able to review this PR before Sunday. Thanks @msoroush for pushing this through to the finish line, and @orbeckst for shepherding it forward.

dotsdl commented:
Went ahead and restarted the build; we'll see if Travis succeeds in shipping to codecov this time.

orbeckst commented:
Nope – somethings is screwy with codecov. For some bizarre reason it lists the coverage for this PR as our project's coverage – see https://codecov.io/gh/alchemistry/alchemlyb. Maybe it gets confused because it comes from the master branch of a fork????

Anyway, we can revisit once this PR is merged.

orbeckst commented:
I manually run the coverage 
```
py.test --cov alchemlyb src/alchemlyb/tests
coverage html
```
with the following result:
```
============================= test session starts ==============================
platform darwin -- Python 3.6.7, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: /Volumes/Data/oliver/Biop/Projects/Methods/FreeEnergy/alchemlyb, inifile:
plugins: xdist-1.20.1, forked-0.2, cov-2.5.1
collected 146 items

src/alchemlyb/tests/test_fep_estimators.py .....................
src/alchemlyb/tests/test_import.py .
src/alchemlyb/tests/test_preprocessing.py ................................
src/alchemlyb/tests/test_ti_estimators.py ...........
src/alchemlyb/tests/test_version.py ..
src/alchemlyb/tests/parsing/test_amber.py ................................................................
src/alchemlyb/tests/parsing/test_gmx.py ..........
src/alchemlyb/tests/parsing/test_gomc.py ..
src/alchemlyb/tests/parsing/test_namd.py ..
src/alchemlyb/tests/parsing/test_util.py .

---------- coverage: platform darwin, python 3.6.7-final-0 -----------
Name                                         Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------------------------
src/alchemlyb/__init__.py                        3      0      0      0   100%
src/alchemlyb/convergence/__init__.py            0      0      0      0   100%
src/alchemlyb/convergence/convergence.py         0      0      0      0   100%
src/alchemlyb/convergence/pade.py                0      0      0      0   100%
src/alchemlyb/estimators/__init__.py             3      0      0      0   100%
src/alchemlyb/estimators/bar_.py                41      0     10      0   100%
src/alchemlyb/estimators/mbar_.py               26      1      4      0    97%
src/alchemlyb/estimators/ti_.py                 31      0      4      0   100%
src/alchemlyb/parsing/__init__.py                0      0      0      0   100%
src/alchemlyb/parsing/amber.py                 235      2     98      4    98%
src/alchemlyb/parsing/gmx.py                   156      1     90      3    98%
src/alchemlyb/parsing/gomc.py                  105      5     48      7    92%
src/alchemlyb/parsing/namd.py                   32      0      8      0   100%
src/alchemlyb/parsing/util.py                   25      4      2      0    85%
src/alchemlyb/preprocessing/__init__.py          4      0      0      0   100%
src/alchemlyb/preprocessing/subsampling.py      43      0     16      0   100%
------------------------------------------------------------------------------
TOTAL                                          704     13    280     14    97%

```

The html pages are attached as 
[coverage.zip](https://github.com/alchemistry/alchemlyb/files/3437302/coverage.zip); note that the current codecov looks similar, e.g., for [alchemlyb/parsing/gomc.py](https://codecov.io/gh/alchemistry/alchemlyb/src/a38807edbca45863936d3eff9d368a964db845f3/src/alchemlyb/parsing/gomc.py).

I hope that helps @dotsdl in his review, even in the absence of codecov.

Based on these results I'll add my own comments in a moment.

orbeckst commented:
The 85% coverage in `parsing/util.py` is harmless – it just needs to be tested under Py2, too.

The lack of testing of a number of except statements in `gomc.py` is an issue, though, it think these lines were copied from the gmx parser without further thinking. I'll make comments in the code.

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    for i, l in enumerate(lambdas):
+        try:
+            u_k[l] = statevec[i]
+        except TypeError:
```
This except clause is not tested. If it does not apply to GOMC then remove it!

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    df = _extract_dataframe(filename)
+
+    # drop duplicate columns if we (stupidly) have them
+    df = df.iloc[:, ~df.columns.duplicated()]
```
I think this can be removed, see recent PR #87 – but double check!

msoroush replied:
In master this line has been removed in gmx.py, but in this PR, it exist still!

orbeckst replied:
Yes, because @dotsdl removed it in PR #87.

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+    for i, l in enumerate(cols):
+        try:
+            dHdl[l] = statevec[i]
+        except TypeError:
```
This except clause is not tested. If it does not apply to GOMC then remove it!

orbeckst commented on src/alchemlyb/parsing/gomc.py:
```diff
@@ -0,0 +1,211 @@
+            # should catch non-numeric lines so we don't proceed in parsing
+            # here
+            if line.startswith('#'):
+                continue
```
This line is not tested.

Can you add a datafile to your dataset that triggers this?

orbeckst replied:
Comments that start with "should do this" should not be in well-tested code....

msoroush replied:
These comments were carry out from gmx parser

orbeckst replied:
If you don't use it then remove it.

orbeckst replied:
Btw, fair enough, they should also not be in the gmx code ;-)

msoroush replied:
:D. I use it actually. The first line of free energy data would trigger it.

orbeckst commented:
Looking good:
```
(alchemlyb) yngvi:alchemlyb oliver$ py.test --cov alchemlyb src/alchemlyb/tests
============================================================ test session starts ============================================================
platform darwin -- Python 3.6.7, pytest-3.2.3, py-1.4.34, pluggy-0.4.0
rootdir: /Volumes/Data/oliver/Biop/Projects/Methods/FreeEnergy/alchemlyb, inifile:
plugins: xdist-1.20.1, forked-0.2, cov-2.5.1
collected 146 items

src/alchemlyb/tests/test_fep_estimators.py .....................
src/alchemlyb/tests/test_import.py .
src/alchemlyb/tests/test_preprocessing.py ................................
src/alchemlyb/tests/test_ti_estimators.py ...........
src/alchemlyb/tests/test_version.py ..
src/alchemlyb/tests/parsing/test_amber.py ................................................................
src/alchemlyb/tests/parsing/test_gmx.py ..........
src/alchemlyb/tests/parsing/test_gomc.py ..
src/alchemlyb/tests/parsing/test_namd.py ..
src/alchemlyb/tests/parsing/test_util.py .

---------- coverage: platform darwin, python 3.6.7-final-0 -----------
Name                                         Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------------------------
src/alchemlyb/__init__.py                        3      0      0      0   100%
src/alchemlyb/convergence/__init__.py            0      0      0      0   100%
src/alchemlyb/convergence/convergence.py         0      0      0      0   100%
src/alchemlyb/convergence/pade.py                0      0      0      0   100%
src/alchemlyb/estimators/__init__.py             3      0      0      0   100%
src/alchemlyb/estimators/bar_.py                41      0     10      0   100%
src/alchemlyb/estimators/mbar_.py               26      1      4      0    97%
src/alchemlyb/estimators/ti_.py                 31      0      4      0   100%
src/alchemlyb/parsing/__init__.py                0      0      0      0   100%
src/alchemlyb/parsing/amber.py                 235      2     98      4    98%
src/alchemlyb/parsing/gmx.py                   156      1     90      3    98%
src/alchemlyb/parsing/gomc.py                   98      1     48      7    95%
src/alchemlyb/parsing/namd.py                   32      0      8      0   100%
src/alchemlyb/parsing/util.py                   25      4      2      0    85%
src/alchemlyb/preprocessing/__init__.py          4      0      0      0   100%
src/alchemlyb/preprocessing/subsampling.py      43      0     16      0   100%
------------------------------------------------------------------------------
TOTAL                                          697      9    280     14    98%


======================================================= 146 passed in 142.37 seconds ========================================================
```

dotsdl commented:
I've made some small changes, including adding `-lambda` to index names for the `u_nk` parser.

One note as an optimization: it might speed up dataframe parsing to use the `pandas.read_csv` parser for the data rows as we did for Gromacs, since this parser can be quite a bit faster than a parser done at the Python interpreter level (it is written in C). This implementation is great as a first pass, however; we are happy to have it.

Thanks @msoroush for sticking it out through this process!

@orbeckst, looks like we lost a little test coverage(?), but it's not clear to me where it could really be improved on this file. My local test run shows it has 95% coverage. Please merge when satsified; thanks!

orbeckst commented:
Thanks. Because PR #85 was merged here and has not been merged into master, this PR has to wait until PR #85 is officially merged.

orbeckst commented:
I also think we'll have to live with the coverage. I'll merge once PR #85 has been merged.

Thanks everyone!

msoroush commented:
Thank you @dotsdl @orbeckst for all your help.

orbeckst commented:
Congratulations @msoroush , your first PR in alchemlyb was merged. Thank you!
